bowling OOP版作成

reckyy opened the pull request:
ご確認の程、よろしくお願いいたします！

reckyy committed:
📃 chore: bowlint_object.rb作成

reckyy committed:
⚡️ feat: Shotクラス作成

reckyy committed:
⚡️ feat: Frameクラス作成

reckyy committed:
⚡️ feat: Gameクラス作成

reckyy committed:
⚡️ feat: bowling_object.rb実装

reckyy committed:
🐛 fix: 色々修正(Rubocopなど)

reckyy committed:
💯 test: rubocop

reckyy committed:
💯 test: rubocop

reckyy committed:
⚡️ feat: とりあえず動くbowlingプログラム作成

reckyy committed:
🐛 fix: initial_number_of_frame10が少ないバグを修正

Xの数を数えてそれを引けばframe10の始まるインデックスを求められると考えていたが、frame10でストライクを出していた場合、その数も数えてしまい値が異なるバグが発生。なので、別途frame10までXの数を数えるメソッドを作成

reckyy committed:
🐛 fix: フレームの2投目で10本倒したとき、ストライクになってしまうのを修正

reckyy committed:
🐛 fix: frameのinitializeを少し変更

もし10フレーム目が2投で終わったとき、shot3がnilになってしまっていた。なのでキーワード引数で0にするのではなく、nilガードを用いて値を代入に変更

reckyy committed:
🍀 refactor: separate_to_frameの内部処理を分割し、メソッド化

reckyy committed:
📃 chore: game.rbのメソッドの並びを修正

reckyy committed:
🍀 refactor: 変数名変更

scoreメソッドに変数scoreが存在していたため

reckyy committed:
💯 test: rubocop

reckyy committed:
🐛 fix: メソッド名を動詞に統一

reckyy committed:
🐛 fix: requireは一番上

reckyy committed:
💯 test: rubocop

betachelsea commented on 06.bowling_object/bowling_object.rb:
```diff
@@ -0,0 +1,7 @@
+require_relative 'game'
+
+game = Game.new
+game.separate_to_frame
```
[MUST] この `separate_to_frame` メソッドは、game のインスタンス自身に影響を与えるだけで、戻り値の活用はなさそうですね。その場合、Gameクラス内で処理を済ませてしまった方が良いかと思います。この処理をしないとボウリングゲームに関する処理が成り立たないのであれば、 Game の initialize メソッドで呼び出してみましょう 💡  また、クラスの外で呼び出さない内部のメソッドは、private扱いにして隠蔽しておくと、不用意に利用されずに済んで安心です。privateについてはお手元に「Ruby超入門」の冊子があれば、再チェックしてみてください！ また、プログラム全体で private の件に関して見直しをしてみてください 🙏

reckyy replied:
[🍀 refactor: separate_to_frameをinitialize時に実行](https://github.com/reckyy/ruby-practices/pull/12/commits/35f9d762853aa83a66a7a14841cdfa40d35a4de1)

betachelsea commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,38 @@
+    shot3 ||= 0
+    @shot1 = shot1
+    @shot2 = shot2
+    @shot3 = shot3
```
[nits] 次のように書くと、`shot3 ||= 0` の行は無くても良さそうです 💡 

```ruby
@shot3 = shot3 || 0
```

reckyy replied:
確かにそうでした。修正しました！
[🍀 refactor: Frame#initializeを修正](https://github.com/reckyy/ruby-practices/pull/12/commits/bc1efac685b966813a6ea6de0a40104c63f924b7)

betachelsea commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,38 @@
+  end
+
+  def strike?
+    @shot1 == 10
```
[nits] 10という値には意味がありそうなので、定数にしておくとより良さそうだな〜と思いました 👀

reckyy replied:
ご指摘の通り、10は全てピンが倒れている状態を示唆するものなので、Shotクラスで`MAX_PINS`として定数化しました！😭
[🍀 refactor: 全ファイルの「10」を定数MAX_PINSに変更](https://github.com/reckyy/ruby-practices/pull/12/commits/487504ed72f12155926f2bbf9256a31616c76d03)
[🐛 fix: MAX_PINSの定義をグローバルスコープに変更](https://github.com/reckyy/ruby-practices/pull/12/commits/4eddc5e4e542ad420ee7f8c346d2d79692ec9ea2)

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,78 @@
+  def separate_to_frame
+    scores = ARGV[0].split(',')
+    shots = []
+    initial_number_of_frame10 = calc_frame9_of_number(scores)
```
[IMO] `calc_frame9_of_number` の戻り値が `initial_number_of_frame10` という変数に代入されているということで、同一の意味合いを示すものと思いますが、「frame9」と「frame10」と別物を扱っているように思えて少し難しく感じました。次のような感じにするのはどうでしょうか。

```ruby
index_of_frame10 = calc_index_of_frame10(scores)
```


(自分用にメモ:  `calc_frame9_of_number` → 10フレーム目の計算をするためのindex値を事前に計算するためのメソッド。 )

betachelsea replied:
[nits] また、`deal_frame10` メソッドに 「initial_number_of_frame10」渡していますが、1回計算したら変わらないものであれば、計算結果をGameクラスのインスタンス変数として保持しておけば、引数として渡さなくても使いまわせて便利かも、と思いました 💡

reckyy replied:
> [IMO] calc_frame9_of_number の戻り値が initial_number_of_frame10 という変数に代入されているということで、同一の意味合いを示すものと思いますが、「frame9」と「frame10」と別物を扱っているように思えて少し難しく感じました。次のような感じにするのはどうでしょうか。

おっしゃる通りです。中身は同じなのに、結果の変数名と求めるメソッドの対象名が異なっており混乱を招く内容でした。申し訳ございません。
ご提示いただいた通り修正いたしました。
[🍀 refactor: 10フレーム目のインデックス = index_of_frame10に統一](https://github.com/reckyy/ruby-practices/pull/12/commits/ca1b31bc2603104aabd343281be2bc91fbf394fa)

> [nits] また、deal_frame10 メソッドに 「initial_number_of_frame10」渡していますが、1回計算したら変わらないものであれば、計算結果をGameクラスのインスタンス変数として保持しておけば、引数として渡さなくても使いまわせて便利かも、と思いました 💡

~@index_of_frame10というインスタンス変数にすることで、`calc_index_of_frame10`メソッドの引数も省略し、直接インスタンス変数に計算結果を代入しようとも考えましたが、それは可読性が悪くなると思いやめておきました。~
最初から、デフォルトの18(ストライクが一度も出なかったら、18回目に10フレーム目が始まるため)を`@index_of_frame10`に代入しておき、全スコアをフレームごとに分ける処理(separate_to_frameメソッド)の際に`@index_of_frame10`を計算するようにしました。
[🐛 fix: index_of_frame10をインスタンス変数として保持](https://github.com/reckyy/ruby-practices/pull/12/commits/ac5bacff2dce2d811ed0a49cccb3d1e106a32933)

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,78 @@
+  end
+
+  def deal_frame10(scores, initial_number_of_frame10)
+    tenth_frame_scores = scores[initial_number_of_frame10..].map { |s| Shot.new(s).pins }
```
[IMO] `Shot.new(s).pins` という使い方だと、new したオブジェクトを単純に変換メソッドの利用のためだけに使って捨てているのは、オブジェクトの活用状況としてはやや薄いなと感じます 🤔  
とはいえ、処理手段的には、Frame.newへの引数に Shotオブジェクトを渡す意義もやや見出しにくいので、数値(Integer)にするのも現状はやむを得ないか、とも思います。もしレビュー後の修正中に、`Shot.new` したオブジェクトを使いまわせる良い方法を思いついたら、取り入れてみてください 💡

reckyy replied:
確かにそうですよね。。自分でもなんか今一つピンと来ていません。。
`create_frame_by_shot`にも組み込めなさそうなので 😭 
何か思いついたらコメントします！

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,78 @@
+  end
+
+  def create_frame_by_shot(shot, shots)
+    if shot.symbol == 'X'
```
[MUST] プログラム中で `X` が3箇所使われていますね。もし仕様が変わって「これから小文字の `x` を使います」と言われたら、3箇所とも修正しなければならず、手間になってしまいそうだなと思いました。
- 定数化して、↑のような仕様変更を伝えられてもすぐ対応できるようにしてください。
- また、オブジェクト指向としては、`X` の存在をしっているのはShotクラスだけ、というカタチになると、Gameクラスが `X` という表現のことを覚えてなくても良いということになり、便利になりそうだと思います。

reckyy replied:
2回以上その変数が使われていたら定数化すべき（柔軟度を保つ）でした。。申し訳ございません。
プログラム中の`'X'`を`shot.rb`で`STRIKE_SYMBOL`として定義しました。
[🍀 refactor: プログラム中の'X'を定数STRIKE_SYMBOLに変更](https://github.com/reckyy/ruby-practices/pull/12/commits/6cb55b9dd14cdf78b56b195f0a6d1e38fa12f477)

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,78 @@
+      next_frame = @frames[i + 1]
+      next_to_frame = @frames[i + 2]
+      if i < 9
+        frame_score = if @frames[i].shot1 == 10 # strike
```
[MUST] Frameには `strike?` メソッドがあるようなので、活用しましょう 💡

reckyy replied:
自身で作成したにもかかわらず、非OOP版のプログラムの仕様のままにしてしまっていました。申し訳ございません。
`strike?`メソッドの仕様と合わせて、`spare?`メソッドも併せて使用する形に変更しました。
[🍀 refactor: ストライクかスペアかの判断をFrame#strike?, Frame#spare?を使用](https://github.com/reckyy/ruby-practices/pull/12/commits/aae2fd02abcf9a31a24958b5dc51c6d479fc4068)

betachelsea reviewed (commented):
拝見しました〜 ✍️

reckyy committed:
🍀 refactor: Frame#initializeを修正

@shot3への代入時に直接nil || 0を実行

reckyy committed:
🍀 refactor: 全ファイルの「10」を定数MAX_PINSに変更

定数の定義は、Shotクラスで行った。

reckyy committed:
🐛 fix: MAX_PINSの定義をグローバルスコープに変更

reckyy committed:
🍀 refactor: separate_to_frameをinitialize時に実行

reckyy committed:
🍀 refactor: プログラム中の'X'を定数STRIKE_SYMBOLに変更

reckyy committed:
🍀 refactor: ストライクかスペアかの判断をFrame#strike?, Frame#spare?を使用

直接検証してしまっていた。メソッドを作成していたにもかかわらず。

reckyy committed:
🍀 refactor: 10フレーム目のインデックス = index_of_frame10に統一

reckyy committed:
🐛 fix: index_of_frame10をインスタンス変数として保持

最初は普通のローカル変数としていたが、一度計算した結果が不変であることからインスタンス変数に。

reckyy committed:
🐛 fix: Frame#sumをFrame#scoreに変更

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,79 @@
+    end
+  end
+
+  def separate_to_frame
```
Gameの中だけで扱うメソッドになりましたね ✨ 

[MUST] 他の場所から誤って呼び出されないように、privateメソッドにしておきましょう！

reckyy replied:
前回ご指摘いただいていたにもかかわらず、未対応で申し訳ございません。
`game.rb`の中で外部で使用しないメソッドをprivateにしました。
[fix: クラス内のみで使うメソッドをprivateに](https://github.com/reckyy/ruby-practices/pull/12/commits/aee3ad58ae9cc92936be6e24742e0c2cda8efc42)

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,79 @@
+class Game
+  def initialize
+    @frames = []
+    @index_of_frame10 = 18
```
[MUST] このように書かれていると、定数にしたくなるのですが、 `calc_index_of_frame10(scores)` で編集される値なので、インスタンス変数である必要があるのですね。`calc_index_of_frame10(scores)` はどちらにせよ`separate_to_frame` メソッドの序盤で呼び出されるので、この辺りを切り出して initialize で済ませることはできないでしょうか？  👀 　

[IMO] 最初の処理だけ用事があるメソッド群を、入れ子で呼び出すと追うのが大変なので、可能ならば平坦に並べておいた方がチェックしやすいように思います。(この辺りはレビュアーの好みもあるかもしれませんが… 🙇  )

reckyy replied:
> [MUST] このように書かれていると、定数にしたくなるのですが、 calc_index_of_frame10(scores) で編集される値なので、インスタンス変数である必要があるのですね。calc_index_of_frame10(scores) はどちらにせよseparate_to_frame メソッドの序盤で呼び出されるので、この辺りを切り出して initialize で済ませることはできないでしょうか？ 👀

~これは修正の時に initialize で済まそうとも考えたのですが、そうなると scores をインスタンス変数にしないといけないから不要なインスタンス変数が増えると思い据え置きにしました。~
ですが今気づきました。普通に initialize 内でscoresを取得してそれを引数で渡せばよかった話でした。
[🐛 fix: 2つのメソッドをinitializeで行うように変更](https://github.com/reckyy/ruby-practices/pull/12/commits/28e8612a8eefb57270dbd5e03021f8e79a211059)

> [IMO] 最初の処理だけ用事があるメソッド群を、入れ子で呼び出すと追うのが大変なので、可能ならば平坦に並べておいた方がチェックしやすいように思います。(この辺りはレビュアーの好みもあるかもしれませんが… 🙇 )

平坦 = メソッドの使用順、と解釈しました。
[🍀 refactor: メソッドを使用順に並び替え](https://github.com/reckyy/ruby-practices/pull/12/commits/c119512eb063c9ead3b9aee8c5b86d33409371f9)

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,79 @@
+  end
+
+  def deal_frame10(scores)
+    tenth_frame_scores = scores[@index_of_frame10..].map { |s| Shot.new(s).pins }
```
> 自分でもなんか今一つピンと来ていません。。

[IMO] `Shot.new(s).pins` について、考えてみてもらったとのことで、自分からjust ideaですが、スコア文字列の配列 scores を、 Shotの配列として扱ったりとかいうのはアリかも? と思いました。 (あくまで just idea なので、やってみたけどうまくいかなかった…はあるかもしれません 🙇 )

reckyy replied:
ご提案いただきありがとうございます！ 😭 何も思いつかなくて申し訳ないです。。
念のため自身の認識をここに記載します。（この認識のもと修正するため、もし間違いがあればご教示ください。）

> スコア文字列の配列 scores を、 Shotの配列として扱ったりとかいうのはアリかも? と思いました。 

- `scores`を数字(得点)の配列としてではなく、Shotのインスタンスの配列として扱う。
- 得点を計算するときに`pins`メソッドを使用して数字に変換する

~ことにより、deal_frame10メソッド内でインスタンス生成する整合性が取れる、という認識です。~
そもそも `scores`にShotのインスタンスたちを代入するため、`deal_frame10`でインスタンス生成しなくてよくなる、ということでしたね。。失礼いたしました。

この認識のもと進めます！

reckyy replied:
上記の認識のもと、修正しました。
- `scores`をShotクラスのインスタンスの配列として扱う。
- `@frames`にフレームの数字配列を入れるときに、`Shot#pins`で数字に変換。

[🐛 fix: 配列scoresをShotクラスのインスタンスの配列に変更](https://github.com/reckyy/ruby-practices/pull/12/commits/6098086403e31591fb89be111e6ec3db41529c7b)

また、上記の変更と併せて `Shot::STRIKE_SYMBOL`の使用をShotクラスの中のみに限定しました。

**理由**
game.rbの中での「ストライクかどうかの判断」は`Shot#pins == 10`ですべて事足りたから。
[🐛 fix: ストライクかどうかは、pins == 10 に統一](https://github.com/reckyy/ruby-practices/pull/12/commits/2243e674d8689697334f6a4ab46a1db6d572235d)

betachelsea commented on 06.bowling_object/shot.rb:
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+MAX_PINS = 10
+STRIKE_SYMBOL = 'X'
```
[MUST] 定数について、特定のオブジェクトのファイル群の中で扱うときは、classの外に書くと混乱が起きる可能性があるので ( 例… shot.rb で MAX_PINS を定義しているのに、game.rbでも うっかり MAX_PINS を定義しちゃったらどちらを使うのか? → 読み込み順で中身が変わるかも… ) 、classの中に定義してみましょう。そして、他のクラスで呼び出すときは `::Shot::MAX_PINS` といった書き方で呼び出してみましょう。そうすると、他ファイルからもどのクラスに定義された定数なのかはっきりわかり、定義場所を探しやすくなりす。

```ruby
class Shot
  MAX_PINS = 10
  # ...略…
end

#=> ::Shot::MAX_PINS で呼び出せる。
```

- 参考: [変数と定数 \(Ruby 3\.2 リファレンスマニュアル\)](https://docs.ruby-lang.org/ja/latest/doc/spec=2fvariables.html)
  - > あるクラスまたはモジュールで定義された定数を外部から参照するためには`::'演算子を用います。またObjectクラスで定義されている定数(トップレベルの定数と言う)を確実に参照するためには左辺無しの`::'演算子が使えます。

もしくはちゃんと定数を定義する特別なファイルを用意するとかの手段もありますが、フレームワークを使うと解決策をすでに準備してくれてる場合もあります。(Railsだと config/initializers/ 以下に準備する…とか )

reckyy replied:
定数についてご指摘いただきありがとうございます。
グローバルスコープで使用してしまっていました。クラス内で使用し他ファイルでは、どこの定数なのかを明記するように心がけます。
[🐛 fix: Shotクラスの定数のスコープを変更](https://github.com/reckyy/ruby-practices/pull/12/commits/1f2a10d3359c40875e1009711a47cb7a7a9097f2)

betachelsea commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,37 @@
+    score == MAX_PINS && !strike?
+  end
+
+  def double_strike?(next_frame, next_to_frame)
```
[nits] `double_strike?` 、 Frameクラス内からしか呼び出しがないので、private にできそうですね！ 💡

reckyy replied:
[🐛 fix: クラス内のみで使うメソッドをprivateに](https://github.com/reckyy/ruby-practices/pull/12/commits/aee3ad58ae9cc92936be6e24742e0c2cda8efc42)

betachelsea reviewed (commented):
修正ありがとうございます！ 拝見しました〜 ✍️

reckyy committed:
🐛 fix: 2つのメソッドをinitializeで行うように変更

calc_index_of_frame10, separate_to_frameメソッドをinitialize内に移動。

reckyy committed:
🍀 refactor: メソッドを使用順に並び替え

reckyy committed:
🐛 fix: Shotクラスの定数のスコープを変更

グローバルではなく、クラス定数に変更。グローバルだともし複数の同名定数が存在していた場合、どの定数を使用しているかわからなくなるため。

reckyy committed:
🐛 fix: 配列scoresをShotクラスのインスタンスの配列に変更

reckyy committed:
🐛 fix: ストライクかどうかは、pins == 10 に統一

それに伴い、symbolのゲッターメソッドも削除

reckyy committed:
🐛 fix: クラス内のみで使うメソッドをprivateに

reckyy committed:
🐛 fix: require 'debug'削除

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,78 @@
+    scores = ARGV[0].split(',').map { |s| Shot.new(s) }
+    @frames = []
+    @index_of_frame10 = 18
+    calc_index_of_frame10(scores)
```
[nits] 現状、`@index_of_frame10`  の取り扱いが次のような感じですね。

```ruby
@index_of_frame10 = 18
calc_index_of_frame10(scores) # この中で @index_of_frame10 を色々と加工している
```

この書き方を初見で見た時、 `@index_of_frame10` が加工されるものである、という認識を持つのが結構難しく、1年後とかになにもかも忘れた状態でコードをみたら結構戸惑いそうだなと思いました。インスタンス変数 ( `@〜` から始まる変数 ) はいろんなところから呼び出せて便利なのですが、一方で、いろんなところで呼び出せるゆえに色んなところで中身を壊されがちな変数と考え、警戒する必要はあります。 なので、代入したり加工したりといった操作が見えづらい場所にあるケースはできるだけ回避したいと個人的には考えています。

今回の `calc_index_of_frame10(scores)` のメソッドを例にとると、2つの問題があるように思います。
- `@index_of_frame10` が加工されるということが見えづらい
- このメソッドは「べき等性」が無い

このメソッドは1回だけ呼び出す前提のものですね。2回以上実行すると `@index_of_frame10` の値が壊れてしまいます。メソッドはできるだけ「べき等性」(冪等性、何回呼び出しても同じ結果になること)を保っておくとわかりやすくておすすめです。

```ruby
# 代入するのはinitializeメソッドで1回きり、あとは呼び出すだけ
@index_of_frame10 = calc_index_of_frame10(scores)

# ...略...

# これなら何回実行しても同じ数値の結果が得られるはずです
# ※動作確認しておらず、動かなかったらすみませんmm
def calc_index_of_frame10(scores)
  # 18、というのは定数にして直接呼び出してもよいのではないかと。定数名やや適当ですがご容赦ください
  index_of_frame10 = MAX_INDEX_OF_FRAME10
  scores.each_with_index do |s, i|
    index_of_frame10 -= 1 if s.pins == Shot::MAX_PINS
    break if i == index_of_frame10 - 1
  end
  index_of_frame10
end
```

べき等性の観点からいうと、 `separate_to_frame(scores)` も本来は次のようになっているとわかりやすそうなのですが、今回の処理都合の観点から難しいかな、と感じました〜 👀 

```ruby
@frames = separate_to_frame(scores) # 他の箇所で @frames は加工しない
```

betachelsea reviewed (commented):
修正ありがとうございました！確認OKです 👍 
「インスタンス変数の加工」と「べき等性」についてお伝えしておきたく、その点はコメントさせていただきました〜 ✍️ あとは実例を見ていただくのが良いかなと思っており、ぜひメンターの提出物をじっくりチェックして、さらに学びを得ていただきたいです！よろしくお願いします 🙏

reckyy committed:
🐛 fix: calc_index_of_frame10に冪等性を持たせた

現状、@index_of_frame10に18を代入し、一回のみ上記メソッドを行いindex_of_frame10を求めていたが、2回目実行すると値が変わり冪等性がない。かつインスタンス変数を破壊していたことから、非破壊メソッドに変更

reckyy committed:
🐛 fix: @fremasに直接代入する形に変更

deal_frame10のshot3代入方法も少し変更している。nilに対しpinsを実行するためエラーが生じたため。

reckyy committed:
🐛 fix: ストライクが正常に判断されていなかった問題を修正

betachelsea commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,39 @@
+  def initialize(shot1, shot2, shot3: 0)
+    @shot1 = shot1
+    @shot2 = shot2
+    @shot3 = shot3
```
[nits] ここの行が、 a73c4061afd9f09cfa のcommitで先祖返りしたようですね 👀

reckyy replied:
そうです！
結局nilではなく最初から０を代入して、10フレーム目のみ `@shot3` に代入すればいいと思いまして。。

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,83 @@
+  def calc_index_of_frame10(scores)
+    index_of_frame10 = Game::MAX_INDEX_OF_FRAME10
+    scores.each_with_index do |s, i|
+      index_of_frame10 -= 1 if s.symbol == Shot::STRIKE_SYMBOL
```
[nits] a9009dfddec7aa4206be33e5 のcommitを拝見していると、 `s.symbol` も結局 Shot クラスのデータを読み出しているので、もしかすると Shotクラスに判定メソッドを生やしても良いかな〜と思いました :eyes:

reckyy replied:
おっしゃる通りです。一応定義してみたのですが、どうでしょうか。 
メソッド名は、Frameクラスのものと同様になってしまいましたが。。
```ruby
def strike?
  @symbol == STRIKE_SYMBOL
end
```

betachelsea replied:
強いて言えばですが、ストライクかどうかというのはShot自身は知ることはできないので ( 1投目で10点ならストライクだが、2投目で10点ならスペアである )、 `strike` というキーワードのないカタチでの表現を模索しても良さそうだと思いました :eyes: 

- is_full_score?
- is_perfect?
- is_perfect_symbol?

(なんとなく `is_a?` 系の命名を参考にして `is_`  はついてても良いかなと思いました )

[Object\#is\_a? \(Ruby 3\.2 リファレンスマニュアル\)](https://docs.ruby-lang.org/ja/latest/method/Object/i/is_a=3f.html)

reckyy replied:
コメントありがとうございます！
> ストライクかどうかというのはShot自身は知ることはできないので

確かにそちらは懸念点ではありました。
rubocopの指摘に引っかかったので、`is`はなしのメソッド名に変更しました。
[🍀 refactor: Shot#strike?をShot#perfict_symbol?に改名](https://github.com/reckyy/ruby-practices/pull/12/commits/065a4484b0f4a7dd1402d9581a02ae6bc5440d01)

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,83 @@
+    frames = []
+    scores.each_with_index do |s, i|
+      if i == @index_of_frame10
+        deal_frame10(scores, frames)
```
[IMO] `frames` の配列をあらゆるところに渡していくのが、ちょっと大変そうですね 🤔 
戻り値で足したい frames のデータを返してあげるのはどうでしょうか。(すみません、仕組み上、可能かどうかの検証はできていないのですが… 🙏 )

```ruby
add_frames = deal_frame10(scores) #=> [#<Frame:...>, #<Frame:....>,...]
frames.concat(add_frames) # 配列と配列を統合するメソッド
```

参考: [Array\#concat \(Ruby 3\.2 リファレンスマニュアル\)](https://docs.ruby-lang.org/ja/latest/method/Array/i/concat.html)

`create_frame_by_shot` メソッドでも、同じことができないか気になります。

reckyy replied:
コメント、更に`concat`メソッドの共有もありがとうございます！

`create_frame_by_shot`メソッドは内部でframesにFrameクラスのインスタンスを追加しています。
なので、`frames`への追加は`push`メソッドで統一しました。
もしこの修正が @betachelsea さんの意図を汲み取れていなければ、ご教示いただけますと幸いです 🙇 
（そもそもの`create_frame_by_shot`メソッドの挙動を変えて欲しかったなど。)

[🍀 refactor: framesの配列を渡す回数を減らした。](https://github.com/reckyy/ruby-practices/pull/12/commits/dd168ddbe84decb88c878a7ae7485190fcf46fa4)

betachelsea reviewed (commented):
修正ありがとうございます！ `calc_index_of_frame10(scores)` が何度実行しても結果が変わらない(引数によってのみ変化する)状態になり、冪等性を獲得しましたね！

他に修正いただいた点について、少し気になる点があったので、コメントしました〜 ✍️

reckyy committed:
🍀 refactor: Shotクラスにもstrike?メソッド実装

reckyy committed:
🍀 refactor: framesの配列を渡す回数を減らした。

ついでにadd_framesメソッドも削除。直接Array#pushで対処。

reckyy commented:
こちらこそご確認いただきありがとうございます！
修正いたしましたので、ご確認の程よろしくお願いいたします。

betachelsea commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,78 @@
+
+  def create_frame_by_shot(shot, shots, frames)
+    if shot.strike?
+      frames.push(Frame.new(shot.pins, 0))
```
> もしこの修正が @betachelsea さんの意図を汲み取れていなければ、ご教示いただけますと幸いです 🙇
> （そもそものcreate_frame_by_shotメソッドの挙動を変えて欲しかったなど。)

[IMO] そうですね、`create_frame_by_shot` について、ちょっと違ったので共有させてください！ 🙏 
( `deal_frame10` については 指摘意図に沿った修正をしてくださったと思います！ )

`create_frame_by_shot` は、まだ引数に frames が存在して、そこにpushするようなカタチになっていますね。

自分は、framesへの加工処理( `frames.push(...)` )がメソッド内に隠れてしまっていると、framesに何が起こっているか分かりづらくなるため、それを `separate_to_frame` 上でのみ実施されたら良いのでは、と考えました。
→  `frames.push` の操作が `create_frame_by_shot` では実施されない(= 引数でframesを渡す必要が無い) カタチをイメージしておりました。

ただ、あくまでイメージしていただけで、実際にそう修正できるかは未トライですので、仕組み上NGということはありうるかも…? と思っております 🙇

reckyy replied:
frames.pushを行うのは、`separate_to_frame`メソッド内のみ、ということ、今理解しました。
申し訳ございません。
ご指摘通り、`create_frame_by_shot`でもフレームの値を返し、もし情報が入っていたらpushする、という形に変更しました。
[🐛 fix: frames.pushをseparate_to_frameメソッド内のみに変更](https://github.com/reckyy/ruby-practices/pull/12/commits/d5a366ab70aa5d2a962f6813eb3733d989ba28b3)

betachelsea replied:
`frames` の扱いについて、わかりやすくなったと思います！ありがとうございます 👍

betachelsea reviewed (commented):
修正ありがとうございます！コメントしました ✍️

reckyy committed:
🍀 refactor: Shot#strike?をShot#perfict_symbol?に改名

reckyy committed:
🐛 fix: frames.pushをseparate_to_frameメソッド内のみに変更
